air/src/utils.rs

```diff
@@ -1,8 +1,172 @@
 use alloc::vec::Vec;
-use core::ops::Range;
+use core::ops::{Add, AddAssign, Index, IndexMut, Mul, Range, Rem, Shr, Sub};
 
 use super::FieldElement;
-use vm_core::utils::range as create_range;
+use vm_core::{utils::range as create_range, Felt};
+
+// NEWTYPE WRAPPERS
+// ================================================================================================
+
+/// A newtype wrapper around a usize value representing a step in the execution trace.
+#[derive(Debug, Copy, Clone)]
+pub struct Step(u32);
+
+// FROM STEP
+
+impl From<Step> for u32 {
+    fn from(step: Step) -> u32 {
+        step.0
+    }
+}
+
+impl From<Step> for i32 {
+    fn from(step: Step) -> i32 {
+        step.0 as i32
+    }
+}
+
+impl From<Step> for usize {
+    fn from(step: Step) -> usize {
+        step.0 as usize
+    }
+}
+
+impl From<Step> for Felt {
+    fn from(step: Step) -> Felt {
+        Felt::from(step.0)
+    }
+}
+
+// INTO STEP
+
+impl From<usize> for Step {
+    fn from(value: usize) -> Step {
+        Step(value as u32)
+    }
+}
+
+impl From<u32> for Step {
+    fn from(value: u32) -> Step {
+        Step(value)
+    }
+}
+
+impl From<i32> for Step {
+    fn from(value: i32) -> Step {
+        Step(value as u32)
+    }
+}
+
+// STEP OPS
+
+impl Sub<usize> for Step {
+    type Output = Step;
+
+    fn sub(self, rhs: usize) -> Self::Output {
+        Step(self.0 - rhs as u32)
+    }
+}
+
+impl Sub<Step> for usize {
+    type Output = Step;
+
+    fn sub(self, rhs: Step) -> Self::Output {
+        (self - rhs.0 as usize).into()
+    }
+}
+
+impl Sub<Step> for Step {
+    type Output = Step;
+
+    fn sub(self, rhs: Step) -> Self::Output {
+        Step(self.0 - rhs.0)
+    }
+}
+
+impl Sub<Step> for u32 {
+    type Output = Step;
+
+    fn sub(self, rhs: Step) -> Self::Output {
+        Step(self - rhs.0)
+    }
+}
+
+impl Add<usize> for Step {
+    type Output = Step;
+
+    fn add(self, rhs: usize) -> Self::Output {
+        Step(self.0 + rhs as u32)
+    }
+}
+
+impl Add<Step> for u32 {
+    type Output = Step;
+
+    fn add(self, rhs: Step) -> Self::Output {
+        Step(self + rhs.0)
+    }
+}
+
+impl AddAssign<usize> for Step {
+    fn add_assign(&mut self, rhs: usize) {
+        self.0 += rhs as u32;
```

Comment:
We should use `rhs.into()` (to use our definition of `usize -> Step` that panics when value doesn't fit in `u32`).

Comment applies everywhere that we use `usize as u32`

Reply:
I believe we still have some `usize as u32` conversions, e.g. in the `impl Sub<usize> for RowIndex` implementation